### PR TITLE
SCIM-scoped API key

### PR DESCRIPTION
**Pull request by noahmoss**
* Adds `:scope` to API keys and adds the ability to create a `:scim`-scoped key. Also drops the NOT NULL constraint on `api_key.user_id`
* Adds `+scim-auth` ring middleware which checks the API key passed in the `Authentication` header
* Creates a basic enterprise namespace structure for SCIM

### CODE REVIEW

**Comment by noahmoss** [comment]
* **#42086** <a href="https://app.graphite.dev/github/pr/metabase/metabase/42086?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @noahmoss and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: test/metabase/api/automagic_dashboards_test.clj [hunk 1]
```diff
@@ -90,7 +90,7 @@
 (deftest metric-xray-test
   (testing "GET /api/automagic-dashboards/metric/:id"
     (t2.with-temp/with-temp [LegacyMetric {metric-id :id} {:table_id   (mt/id :venues)
-                                                     :definition {:query {:aggregation ["count"]}}}]
+                                                           :definition {:query {:aggregation ["count"]}}}]
```
Comment (noahmoss) [inline test/metabase/api/automagic_dashboards_test.clj:93]: Unrelated indentation fixes

File: enterprise/backend/src/metabase_enterprise/scim/v2/api.clj [hunk 2]
```diff
@@ -0,0 +1,14 @@
+(ns metabase-enterprise.scim.v2.api
+  "/api/ee/scim/v2/ endpoints. These are the endpoints which implement the SCIM protocol, as opposed to SCIM
+  configuration endpoints which are in `metabase-enterprise.scim.api`."
+  (:require
+   [compojure.core :refer [GET]]
+   [metabase.api.common :as api :refer [defendpoint]]
+   [toucan2.core :as t2]))
+
+(defendpoint GET "/Users"
```
Comment (noahmoss) [inline enterprise/backend/src/metabase_enterprise/scim/v2/api.clj:9]: Not implemented yet for SCIM — just using this as a dummy endpoint to test auth

File: enterprise/backend/src/metabase_enterprise/scim/v2/api.clj [hunk 3]
```diff
@@ -0,0 +1,14 @@
+(ns metabase-enterprise.scim.v2.api
+  "/api/ee/scim/v2/ endpoints. These are the endpoints which implement the SCIM protocol, as opposed to SCIM
+  configuration endpoints which are in `metabase-enterprise.scim.api`."
```
Comment (noahmoss) [inline enterprise/backend/src/metabase_enterprise/scim/v2/api.clj:3]: Wasn't sure about the namespace names and API routes here. The actual SCIM protocol is conventionally at `/scim/v2` but we also have normal APIs for creating/deleting keys, which I've put at `/scim`. Can change this approach if anyone has a better idea.

**Review by johnswanson — APPROVED** [review]
